app/services/badge_service.rb

```diff
@@ -0,0 +1,49 @@
+class BadgeService
+  RULES = %w[all_in_category first_try_success all_in_level].freeze
+  
+  def initialize(test_passage)
+    @test_passage = test_passage
+    @user = test_passage.user
+    @test = test_passage.test
+  end
+
+  def call
+    return unless @test_passage.passed?
+
+    Badge.all.select { |badge| send("assign_#{badge.rule}?", badge.condition) }
+  end
+
+  private
+
+  def assign_all_in_category?(category)
+    return unless @test.category.title == category
+
+    tests = Test.by_category(category).ids
+    check(tests, category)
+  end
+
+  def assign_first_try_success?(condition)
+    @user.tests.where(id: @test.id).count == 1
+  end
+
+  def assign_all_in_level?(level)
+    return unless @test.level.to_i == level.to_i
+
+    tests = Test.by_level(level).ids
+    check(tests, level)
+  end
+
+  def check(tests, condition)
+    badge = @user.badges.where(condition: condition).last
+
+    if badge.present?
+      last_badge = @user.user_badges.where(badge_id: badge.id).last
```

Comment:
```suggestion
      last_assignment = @user.user_badges.where(badge_id: badge.id).last
```
это ведь не бейдж, а "выдача бейджа"
и еще бы я здесь добавил сортировку по `created_at`, чтобы быть уверенным в результате